hexkit/providers/s3/provider.py

```diff
@@ -0,0 +1,808 @@
+
+        if await self.does_object_exist(bucket_id=bucket_id, object_id=object_id):
+            raise self.ObjectAlreadyExistsError(
+                bucket_id=bucket_id, object_id=object_id
+            )
+
+    async def _get_object_upload_url(
+        self,
+        *,
+        bucket_id: str,
+        object_id: str,
+        expires_after: int = 86400,
+        max_upload_size: Optional[int] = None,
+    ) -> PresignedPostURL:
+        """Generates and returns an HTTP URL to upload a new file object with the given
+        id (`object_id`) to the bucket with the specified id (`bucket_id`).
+        You may also specify a custom expiry duration in seconds (`expires_after`) and
+        a maximum size (bytes) for uploads (`max_upload_size`).
+        """
+
+        await self._assert_object_not_exists(bucket_id=bucket_id, object_id=object_id)
+
+        conditions = (
+            []
+            if max_upload_size is None
+            else [
+                # set upload size limit:
+                ["content-length-range", 0, max_upload_size],
+            ]
+        )
+
+        try:
+            presigned_url = await asyncio.to_thread(
+                self._client.generate_presigned_post,
+                Bucket=bucket_id,
+                Key=object_id,
+                Conditions=conditions,
+                ExpiresIn=expires_after,
+            )
+        except botocore.exceptions.ClientError as error:
+            raise self._translate_s3_client_errors(
+                error, bucket_id=bucket_id, object_id=object_id
+            ) from error
+
+        return PresignedPostURL(
+            url=presigned_url["url"], fields=presigned_url["fields"]
+        )
+
+    async def _list_mulitpart_upload_for_object(
+        self, *, bucket_id: str, object_id: str
+    ) -> list[str]:
+        """Lists all active multi-part upload for the given object. Returns a list of
+        their IDs.
+
+        (S3 allows multiple ongoing multi-part uploads.)
+        """
+
+        try:
+            uploads_info = await asyncio.to_thread(
+                self._client.list_multipart_uploads,
+                Bucket=bucket_id,
+            )
+        except botocore.exceptions.ClientError as error:
+            raise self._translate_s3_client_errors(
+                error, bucket_id=bucket_id, object_id=object_id
+            ) from error
+
+        if "Uploads" in uploads_info:
+            upload_list = uploads_info["Uploads"]
+            return [
+                upload["UploadId"]
+                for upload in upload_list
+                if upload["Key"] == object_id
+            ]
+
+        return []
+
+    async def _assert_no_multipart_upload(self, *, bucket_id: str, object_id: str):
+        """Ensure that there are no active multi-part uploads for the given object."""
+
+        upload_ids = await self._list_mulitpart_upload_for_object(
+            bucket_id=bucket_id, object_id=object_id
+        )
+        if len(upload_ids) > 0:
```

Comment:
Actually, this falsiness/truthiness is one of the behaviors of python which I don't find particulary helpful. `if upload_ids` could mean a lot of things (including `if upload_ids is not None` and you always need contect to understand for which of these senarios you are checking. That's why I like if checks like this are explicit. Don't you agree?

Reply:
I actually like how Python (but not JavaScript) does it. In cases like this one I don't mind whether upload_ids is None or an empty list, in both cases I don't need to deal with it. And then "if upload_ids:" is more readable and expresses exactly what I want. And since Python *has* this feature, I think it would not be Pythonic to use it.
 
I would argue that
 - Explicit is better than implicit.
 does not trump these:
- Beautiful is better than ugly.
- Simple is better than complex.
- Readability counts.

Reply:
I have added this question to our style guide issue now.